File: CSharp/RandomNumberGeneration/BlumBlumShub/Utils/MathUtil.cs
namespace RandomNumberGenerators.Utils {
    public static class MathUtil {
        public static long ModPow(long number, long exponent, long modulus) {
            if(modulus == 1) return 0;

            long result = 1;
            number = number % modulus;
            while(exponent > 0) {
                if(exponent % 2 == 1) {
                    result = (result * number) % modulus;
                }
                exponent >>= 1;
                number = (number * number) % modulus;
            }

            return result;
        }
    }
}

File: Lua/utils/mathUtil.lua
-- mathUtil.lua
-- 
-- 

local mathUtil = {}

function mathUtil.modPow(number, exponent, modulus)
	if modulus == 1 then return 0 end
	
	local result = 1
	number = number % modulus
	while exponent > 0 do
		if exponent % 2 == 1 then
			result = (result * number) % modulus
		end
		exponent = exponent >> 1
		number = (number * number) % modulus
	end
	
	return result
end

return mathUtil
